src/tool_operate.c

```diff
@@ -888,6 +892,124 @@ static CURLcode single_transfer(struct GlobalConfig *global,
           }
         }
 
+        /* disallowing simultaneous use of --etag-save and --etag-compare */
+        if(config->etag_save_file && config->etag_compare_file) {
+          warnf(
+            config->global,
+            "Cannot use --etag-save and --etag-compare at the same time\n");
+
+          result = CURLE_UNKNOWN_OPTION;
+          break;
+        }
+
+        /* --etag-save */
+        etag_save = &per->etag_save;
+        etag_save->stream = stdout;
+        etag_save->config = config;
+        if(config->etag_save_file) {
+          /* open file for output: */
+          if(strcmp(config->etag_save_file, "-")) {
+            FILE *newfile = fopen(config->etag_save_file, "wb");
+            if(!newfile) {
+              warnf(
+                config->global,
+                "Failed to open %s\n", config->etag_save_file);
+
+              result = CURLE_WRITE_ERROR;
+              break;
+            }
+            else {
+              etag_save->filename = config->etag_save_file;
+              etag_save->s_isreg = TRUE;
+              etag_save->fopened = TRUE;
+              etag_save->stream = newfile;
+            }
+          }
+          else {
+            /* always use binary mode for protocol header output */
+            set_binmode(etag_save->stream);
+          }
+        }
+
+        /* --etag-compare */
+        if(config->etag_compare_file) {
+          char *etag_from_file = NULL;
+          size_t lgn;
+          char *header;
+          size_t file_size;
+
+          /* open file for reading: */
+          FILE *file = fopen(config->etag_compare_file, FOPEN_READTEXT);
+          if(!file) {
+            warnf(
+              config->global,
+              "Failed to open %s\n", config->etag_compare_file);
+
+            result = CURLE_READ_ERROR;
+            break;
+          }
+
+          /* get file size */
+          fseek(file, 0, SEEK_END);
+          file_size = ftell(file);
+
+          etag_from_file = malloc(file_size + 1);
+          if(!etag_from_file) {
+            warnf(
+              config->global,
+              "Failed to allocate memory for custom etag\n");
+
+            result = CURLE_OUT_OF_MEMORY;
+            break;
+          }
+
+          /*
+           * check if file is empty, if it's not load etag
+           * else continue with empty etag
+           */
+          if(file_size != 0) {
+            fseek(file, 0, SEEK_SET);
+            file2string(&etag_from_file, file);
```

Comment:
file2string allocates memory for the file by itself, no need to allocate the memory before this - and actually this creates a memory leak!

Reply:
thanks, done